Source/Particles/ParticleCreation/FilterCreateTransformFromFAB.H

```diff
@@ -0,0 +1,190 @@
+/* Copyright 2020 Luca Fedeli, Neil Zaim
+ *
+ * This file is part of WarpX.
+ *
+ * License: BSD-3-Clause-LBNL
+ */
+
+#ifndef FILTER_CREATE_TRANSFORM_FROM_FAB_H_
+#define FILTER_CREATE_TRANSFORM_FROM_FAB_H_
+
+/**
+ * \brief Apply a filter on a list of FABs, then create and apply a transform
+ * operation to the particles depending on the output of the filter.
+ *
+ * \tparam N number of particles created in the dst(s) in each cell
+ * \tparam DstTile the dst particle tile type
+ * \tparam FAB the src FAB type
+ * \tparam Index the index type, e.g. unsigned int
+ * \tparam CreateFunc1 the create function type for dst1
+ * \tparam CreateFunc2 the create function type for dst2
+ * \tparam TransFunc the transform function type
+ *
+ * \param[in,out] dst1 the first destination tile
+ * \param[in,out] dst2 the second destination tile
+ * \param[in] box the box where the particles are created
+ * \param[in] src_FAB A FAB defined on box that is used in the transform function
+ * \param[in] mask pointer to the mask: 1 means create, 0 means don't create
+ * \param[in] dst1_index the location at which to starting writing the result to dst1
+ * \param[in] dst2_index the location at which to starting writing the result to dst2
+ * \param[in] create1 callable that defines what will be done for the create step for dst1.
+ * \param[in] create2 callable that defines what will be done for the create step for dst2.
+ * \param[in] transform callable that defines the transformation to apply on dst1 and dst2.
+ *
+ * \return num_added the number of particles that were written to dst1 and dst2.
+ */
+template <int N, typename DstTile, typename FAB, typename Index,
+          typename CreateFunc1, typename CreateFunc2, typename TransFunc,
+          amrex::EnableIf_t<std::is_integral<Index>::value, int> foo = 0>
+Index filterCreateTransformFromFAB (DstTile& dst1, DstTile& dst2, const amrex::Box box,
+                                const FAB *src_FAB, const Index* mask,
+                                const Index dst1_index, const Index dst2_index,
+                                CreateFunc1&& create1, CreateFunc2&& create2,
+                                TransFunc&& transform) noexcept
+{
+    using namespace amrex;
+
+    const auto ncells = box.volume();
+    if (ncells == 0) return 0;
+
+    const auto box_sizex = box.length(0);
+    const auto box_sizey = box.length(1);
+
+    auto & warpx = WarpX::GetInstance();
+    const int level_0 = 0;
+    Geometry const & geom = warpx.Geom(level_0);
+
+    Real xlo_global = geom.ProbLo(0);
+    Real ylo_global = geom.ProbLo(1);
+    Real dx         = geom.CellSize(0);
+    Real dy         = geom.CellSize(1);
```

Comment:
Hmm no I probably got confused with the convention. I think I was writing this using the AMReX doc which I think doesn't have this convention. If you think it's better I can change this (that would add an extra `#if (AMREX_SPACEDIM == 2)` right ?) or at least add a comment specifying that we don't use the usual convention here, in case we want to change this in the future.

Reply:
Either would be fine. And we know this convention (not applied everywhere, and inconsistent with AMReX) is a problem in WarpX, this is an active topic of discussion these days.

Reply:
Ok thanks, I've simply added a comment stating that we use x and y in 2D here.